### ISSUE TITLE
UniMath-edu

### ISSUE THREAD

**Issue by vladimirias** [UniMath/UniMath#77]
I suggest to create a branch of the UniMath called UniMath-edu (or something better) for the version of UniMath and of individual files of UniMath that are oriented towards teaching of UniMath and its style of formalization to mathematicians. There we will keep total2 to be an "Inductive", will avoid destruct (until a good explanation of the "match" to mathematicians can be found) etc.

On the main UniMath we will switch to total2 being a record, use primitive projections and start using canonical structures. Also on the main branch we will start using more advance features of Coq in the newer files (under the condition that the use of any such feature is explicitly described in the README of the corresponding directory!).

**Comment by DanGrayson** [comment]
From "UniMath and its style of formalization" it seems you think there is one style of formalization.  But the two branches will have different styles, and one of them will not be taught.

**Comment by vladimirias** [comment]
I suggest that the earlier files of the UniMath (Foundations, RezkCompletion) will be preserved in the way they are now (modulo the ongoing beautification). The only change will be in uuu.v and it will be with replacing the inductive definition of total2 with the Record definition.

This will enable those who write more advance code to use canonical structures with the total2.

The teaching of the basics of the UniMath will use the version with “Inductive” so that one can “print” various proofs during the demonstrations to show how they are built using the “total2_rect” (whose semantics is easy to explain).

Check the difference here:

****

Set Primitive Projections.

Record test ( T : UU ) ( P : T -> UU ) := test_constr { pr1 : T ; pr2 : P pr1 } .

Arguments pr1 { _ } { _ } _ .
Arguments pr2 { _ } { _ } _ .

Lemma testlemma ( T : UU ) ( P : T -> UU ) ( xp : test T P ) :
  xp = test_constr _ _ ( pr1 xp ) ( pr2 xp ) .
Proof .
  intros .
  induction xp as [ x p ] .
  apply idpath .

Defined.

Print testlemma .

Inductive test' ( T : UU ) ( P : T -> UU ) :=
  test'_constr : forall ( x : T ) ( p : P x ) , test' T P .

Definition pr1' { T : UU } { P : T -> UU } ( xp : test' T P ) : T .
Proof .
  intros . induction xp as [ x p ] . apply x .

Defined.

Definition pr2' { T : UU } { P : T -> UU } ( xp : test' T P ) : P ( pr1' xp ).
Proof .
  intros T P xp . induction xp as [ x p ] . apply p .

Defined.

Lemma testlemma' ( T : UU ) ( P : T -> UU ) ( xp : test' T P ) :
  xp = test'_constr _ _ ( pr1' xp ) ( pr2' xp ) .
Proof .
  intros .
  induction xp as [ x p ] .
  apply idpath .

Defined.

Print testlemma' .

****

Basically, UniMath-edu should probably only contain Foundations and RezkCompletion with everything else being based on the Record approach to total2.

Vladimir.

**Comment by DanGrayson** [comment]
Okay.  Ktheory also can be in edu-mode.

**Comment by DanGrayson** [comment]
Adopting this proposal would lead to a lot of extra work for those who modify any of the earlier
files that are part of UniMath-edu, for any change to those files on the master branch would have
to also be made to the corresponding files on the edu branch.  If a pull request to the master
branch modifies early files and late files, then someone has to submit a modified pull request for
the edu branch that modifies just the early files the same way.

**Comment by vladimirias** [comment]
I think we should put UniMath-edu on hold right now. It is a very serious project and unless one of us wants to invest dome major time and effort into it it is best not to get involved in it now.

This does not mean that I am for relaxing the UniMath style. There is still the requirement of no inductive definitions or records outside of uuu.v and no use of the inductive definitions of the standard library except the ones that have been listed in uuu.v


Similarly there should be no use of Prop or Set.

About Set - we do use it now. Could someone (Benedikt?) prepare a PR which should be very simple in which uuu will be introduced as the lowest universe and as Type (not Set) so that the text at the end of uuu says paths 0 0 : uuu ?

Vlaidmir.

**Comment by benediktahrens** [comment]
On 09/19/2015 01:36 PM, Vladimir Voevodsky wrote:
> 
> About Set - we do use it now. Could someone (Benedikt?) prepare a PR
> which should be very simple in which uuu will be introduced as the
> lowest universe and as Type (not Set) so that the text at the end of
> uuu says paths 0 0 : uuu ?

The datatype "nat" is declared before UUU and UU are introduced, namely
in Coq.Init.Datatypes.
Should we declare our own "nat" living in UUU?

(In this way we might have to reimplement the notation "1" for (S O),
and so on, which, I think, are implemented by an ocaml plugin. I am not
sure about that at all.)

Related, but a bit more globally: currently coqtop loads the Coq
standard prelude (with the "nat" datatype, and some other stuff).
In HoTT, they load coqtop with an option to load their own prelude.
Is that something we would like to have as well?

**Comment by vladimirias** [comment]
> On Sep 20, 2015, at 12:40 PM, Benedikt Ahrens <notifications@github.com> wrote:
> 
> The datatype "nat" is declared before UUU and UU are introduced, namely
> in Coq.Init.Datatypes.
> Should we declare our own "nat" living in UUU?

what happens if we redefine nat in uuu.v as follows:

Definiton nat := ( nat : uuu ) .

**Comment by vladimirias** [comment]
> On Sep 20, 2015, at 12:40 PM, Benedikt Ahrens <notifications@github.com> wrote:
> 
> Related, but a bit more globally: currently coqtop loads the Coq
> standard prelude (with the "nat" datatype, and some other stuff).
> In HoTT, they load coqtop with an option to load their own prelude.
> Is that something we would like to have as well?


I think this should be one of the first in the list of issues that we will create for this year work.

The first step in the work on this issue is to have a look at the current coq.init . I should probably be the main person deciding on which of the features of the coq.init we can (or should at this time) have in the UniMath and which we should not.

If there is at least one component in coq.init that we do not want in UniMath then we should have the UniMath option for the prelude on calling the standard Coq.

**Comment by DanGrayson** [comment]
Inserting ```Definiton nat := ( nat : UUU ) .``` leads to an error later on:

```
"sub/coq/bin/coqc"  -q  -R "UniMath" UniMath  -indices-matter -type-in-type  UniMath/Foundations/hlevel2/hnat
File "./UniMath/Foundations/hlevel2/hnat.v", line 989, characters 77-78:
Error:
In environment
a : nat
b : nat
c : nat
d : nat
The term "a" has type "nat" while it is expected to have type "pr1hSet ?X0".
```

**Comment by DanGrayson** [comment]
> About Set - we do use it now. Could someone (Benedikt?) prepare a PR which
> should be very simple in which uuu will be introduced as the lowest universe
> and as Type (not Set) so that the text at the end of uuu says paths 0 0 : uuu ?

Why not use Set?  In Coq, it is simply the lowest universe level above Prop, and there is
nothing else to distinguish it.

**Comment by DanGrayson** [comment]
... aside from the name, which may confuse users and readers of proofs.  That would be a good justification for avoiding it.

**Comment by DanGrayson** [comment]
To avoid the appearance of "Set" in printed output, we could more simply use
```
Notation uuu:=Set.
```

**Comment by DanGrayson** [comment]
I'd like to continue using ```Record```.  Since it's just an iterated Sigma type, it should be sound to do so, and the coding convenience it offers is great.  Just look at this record I use in ```AbelianGroup.v```:
```
  Record AdequateRelation {X I} (R:I->reln X) (r : hrel (word X)) := 
    make_AdequateRelation {
        base: forall i, r (lhs (R i)) (rhs (R i));
        reflex : forall w, r w w;
        symm : forall v w, r v w -> r w v;
        trans : forall u v w, r u v -> r v w -> r u w;
        left_compat : forall u v w, r v w -> r (word_op u v) (word_op u w);
        right_compat: forall u v w, r u v -> r (word_op u w) (word_op v w);
        left_unit : forall w, r (word_op word_unit w) w;
        right_unit : forall w, r (word_op w word_unit) w;
        assoc : forall u v w, r (word_op (word_op u v) w) (word_op u (word_op v w));
        inverse_compat : forall v w, r v w -> r (word_inv v) (word_inv w);
        left_inverse : forall w, r (word_op (word_inv w) w) word_unit;
        right_inverse: forall w, r (word_op w (word_inv w)) word_unit;
        comm : forall v w, r (word_op v w) (word_op w v)
      }.
```

**Comment by DanGrayson** [comment]
I propose that this issue be closed and that its goals be adopted as part of this issue: https://github.com/UniMath/UniMath/issues/763

Feedback?

**Comment by DanGrayson** [comment]
One reason to do so is that maintaining two branches in perpetuity is just too hard.  Another is that those perusing the system in order learn it want to peruse just one system, and to be assured that their learning will be useful later.
